pennylane/optimize/rotosolve.py

```diff
@@ -76,45 +76,60 @@ class RotosolveOptimizer:
     """
     # pylint: disable=too-few-public-methods
 
-    def step_and_cost(self, objective_fn, x):
+    def step_and_cost(self, objective_fn, *args, **kwargs):
         r"""Update x with one step of the optimizer and return the corresponding objective
         function value prior to the step.
 
         Args:
             objective_fn (function): The objective function for optimization. It should take a
                 sequence of the values ``x`` and a list of the gates ``generators`` as inputs, and
                 return a single value.
-            x (Union[Sequence[float], float]): sequence containing the initial values of the
+            *args (Union[Sequence[float], float]): sequence containing the initial values of the
                 variables to be optimized over or a single float with the initial value
+            **kwargs : Variable length dictionary of keywords for the cost function
 
         Returns:
             tuple: the new variable values :math:`x^{(t+1)}` and the objective function output
                 prior to the step
         """
-        x_new = self.step(objective_fn, x)
+        x_new = self.step(objective_fn, *args, **kwargs)
 
-        return x_new, objective_fn(x)
+        return x_new, objective_fn(*args, **kwargs)
 
-    def step(self, objective_fn, x):
-        r"""Update x with one step of the optimizer.
+    def step(self, objective_fn, *args, **kwargs):
+        r"""Update args with one step of the optimizer.
 
         Args:
             objective_fn (function): The objective function for optimization. It should take a
                 sequence of the values ``x`` and a list of the gates ``generators`` as inputs, and
                 return a single value.
-            x (Union[Sequence[float], float]): sequence containing the initial values of the
-                variables to be optimized over or a single float with the initial value
+            *args (Union[Sequence[float], float]): variable length sequence containing the initial
+                values of the variables to be optimized over or a single float with the initial
+                value
+            **kwargs : Variable length dictionary of keywords for the cost function
 
         Returns:
             array: the new variable values :math:`x^{(t+1)}`
         """
-        x_flat = np.fromiter(_flatten(x), dtype=float)
-        objective_fn_flat = lambda x_flat: objective_fn(unflatten(x_flat, x))
+        before_args = []
+        after_args = list(args[1:])
+        args_new = list(args)
```

Comment:
I think I'm having a bit of trouble following this part. It seems like `before_args` will always be an empty list? And `after_args` will always be a list of all arguments except the first one?

Reply:
Oops! I'll finish that up... maybe I should have some better tests on it?

Reply:
fixed version added

Reply:
Nice! Looks like this should work fine... although it would be nice to add some more comments here as well Partly my fault, since I wrote this quite some time ago, and my commenting here wasn't the best either. :laughing: 

Sorry about that!

Reply:
>Oops! I'll finish that up... maybe I should have some better tests on it?

Yep, this is a sign it needs a test for this specific edge case 🙂